bookkeeper-server/src/main/java/org/apache/bookkeeper/client/RegionAwareEnsemblePlacementPolicy.java

```diff
@@ -134,7 +134,8 @@ public void handleBookiesThatJoined(Set<BookieId> joinedBookies) {
             topology.add(node);
             knownBookies.put(addr, node);
             historyBookies.put(addr, node);
-            String region = getLocalRegion(node);
+            String region = parseBookieRegion(addr);
```

Comment:
Shall we update the `address2Region` in the `onBookieRackChange`. 
Even if the bookie already has been shutdown, we can still update the `address2Region` in the `onBookieRackChange`

Reply:
I think current code has updated address2Region in onBookieRackChange(). If oldRegion == newRegion, no need to update. If oldRegion != newRegion, would put in newRegion.

 https://github.com/apache/bookkeeper/blob/9aad0453f0f17f45ea673e82fb63762156315e8b/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/RegionAwareEnsemblePlacementPolicy.java#L188-L194

Reply:
> I think current code has updated address2Region in onBookieRackChange(). If oldRegion == newRegion, no need to update. If oldRegion != newRegion, would put in newRegion.
> 
> https://github.com/apache/bookkeeper/blob/9aad0453f0f17f45ea673e82fb63762156315e8b/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/RegionAwareEnsemblePlacementPolicy.java#L188-L194

If the line_178 node is null, it does nothing.

Reply:
Is we need to get correct rack of a shutdown bookie ? It seem is enough to update bookie's rack once it restart.

Reply:
I check the code. It's hard to get the rack info when the bookie is shutdown, because we can't resolve the address.

Your solution is good enough.